prow/cmd/branchprotector/protect.go

```diff
@@ -276,11 +279,34 @@ func (p *protector) UpdateBranch(orgName, repo string, branchName string, branch
 		logrus.Infof("%s/%s=%s: already unprotected", orgName, repo, branchName)
 		return nil
 	}
+
+	// The github API currently does not support listing protections for all
+	// branches of a repository. We therefore have to make individual requests
+	// for each branch.
+	currentBP, err := p.client.GetBranchProtection(orgName, repo, branchName)
+	if err != nil {
+		return fmt.Errorf("get current branch protection: %v", err)
+	}
+	var policyBP github.BranchProtection
 	var req *github.BranchProtectionRequest
 	if *bp.Protect {
 		r := makeRequest(*bp)
 		req = &r
+		policyBP = github.BranchProtection{
+			EnforceAdmins: github.EnforceAdmins{
+				Enabled: r.EnforceAdmins != nil && *r.EnforceAdmins,
+			},
+			RequiredPullRequestReviews: r.RequiredPullRequestReviews,
+			RequiredStatusChecks:       r.RequiredStatusChecks,
+			Restrictions:               r.Restrictions,
+		}
 	}
+
+	if req != nil && currentBP != nil && equalBranchProtections(*currentBP, policyBP) {
+		logrus.Debugf("%s/%s=%s: current branch protection matches policy, skipping", orgName, repo, branchName)
+		return nil
+	}
```

Comment:
I'm pretty sure we can also skip if `req == nil && currentBP == nil`. That will prevent unnecessary calls to `RemoveBranchProtection` here: https://github.com/kubernetes/test-infra/blob/021b9b2aea9b9254d38d1a5481bf7ff4108d8cad/prow/cmd/branchprotector/protect.go#L156

Reply:
Thanks for pointing that out. While reviewing the existing unit tests, seems like that particular case, i.e. currently unprotected branch and policy sets protected to false, is already covered:

https://github.com/kubernetes/test-infra/blob/ba215daf8be54c80be6be4c66503b1a5e60d09c9/prow/cmd/branchprotector/protect_test.go#L678-L688

and 

https://github.com/kubernetes/test-infra/blob/ba215daf8be54c80be6be4c66503b1a5e60d09c9/prow/cmd/branchprotector/protect.go#L275-L278

I think that's because the github API's reponse to `GetBranches` has a `protected` field in the payload, so the existing code uses it for that particular optimization.

Please let me know if this is not correct and the `req == nil && currentBP == nil` optimisation is needed .